### ISSUE TITLE
[Question] Using FeatureScaled kernel to perform ARD with GaussianProcess

### ISSUE THREAD

**Issue by fonnesbeck** [tensorflow/probability#741]
If I understand things correctly, one should now be able to perform automatic relevance determination for `GaussianProcess`es using the `FeatureScaled` kernel as follows:

```python
    amp = tf.Variable(dtype(1), name='amplitude')
    obs_var = tf.Variable(dtype(1), name='observation_noise_variance')
    ls = tf.Variable(dtype([1.]*X.shape[1]), name='length_scale')
    m = tf.Variable(dtype(0), name='mean_const')

    K = tfk.FeatureScaled(tfk.MaternThreeHalves(amplitude=amp), 
                        scale_diag=ls)
    gp = tfd.GaussianProcess(mean_fn=lambda x: m,
                            kernel=K,
                            index_points=X,
                            observation_noise_variance=tf.exp(obs_var))
```

However, when I optimize this model, the length scales do not appear to be working independently:

```
<tf.Variable 'length_scale:0' shape=(2,) dtype=float64, numpy=array([1.62942848, 1.62877862])>
```

This does not appear to be merely bad luck, as it works with several different datasets. Am I using this approach improperly?

Running TFP 0.10.0-dev20200122 with TF nightly on Debian with Python 3.7.

**Comment by brianwa84** [comment]
Are you sure the features don't always have the same ideal scale? They do at least appear to be nominally different, maybe the first is moving up very slowly and the second moving down very slowly. I think we'd want to look at more info like learning rate/optimizer, possibly even some samples of the observed gradients.

An obvious hazard I see here is `tf.exp(obs_var)`. Instead of that line I'd suggest using `obs_var = tfp.util.TransformedVariable(dtype(1), bijector=tfb.Exp())`. Otherwise, depending on whether or not the `tf.exp` call is inside or outside the `tf.GradientTape` scope, you might lose the gradient link to the underlying `tf.Variable`. You might try similar for other variables, as generally most of the params to GPs need to be strictly positive (noise variance, amplitude, length_scale).

@csuter @srvasude fyi re GPs and ARD

**Comment by brianwa84** [comment]
Also it's interesting to note that the optimized values are not far from exp(1), I wonder if maybe the noise is dominating any signal?

**Comment by csuter** [comment]
Brian maybe you meant sqrt(exp(1))? This is about 1.648 (exp(1) is just e
~= 2.18).

In any case yeah using transformed variable will avoid losing the gradient
signal and allow the noise to train (assuming, as Brian said, you're
instantiating everything outside the gradient tape). Note that
TransformedVariable is initialized with the *post* transformation value,
whereas initializing a tf.Variable with value 1 and taking the exp gives
you e.

**Comment by brianwa84** [comment]
Ah you're right, it looks close to a different magic number, the golden
ratio, phi
1.618033989

**Comment by fonnesbeck** [comment]
Thanks for the advice. Removing the transformation did not fix the issue, but changing my prior appears to have. For some reason, using an exponential prior kept the lengthscale parameters close in value to one another, but switching to a Lognormal seems to have done the trick. I now get estimates like

    <tf.Variable 'length_scale:0' shape=(2,) dtype=float64, numpy=array([1.30257412 0.7647567 ])>

which is closer to what I would expect, at least. I can't explain why, exactly, but it works!

**Comment by csuter** [comment]
Glad you're up and running again!
